xaitk_saliency/utils/masking.py

```diff
@@ -1,33 +1,179 @@
+    In the mask matrix, values closer to 1 correspond to regions of the image
+    that should *NOT* be occluded.
+    E.g. a 0 in the mask will translate to "blacking out" the corresponding
+    location in the source image.
+
+    We optionally take in a fill-color value as input. This may be either a
+    scalar or per-channel sequence. When no fill is passed, black is used
+    (default absence of color).
+
+    Images output will mirror the input image format. As such, the `fill` value
+    passed must be compatible with the input image channels for broadcasting.
+    For example, a single channel input will not be able to be broadcast
+    against a multi-channel `fill` input. A ValueError will be raised by the
+    underlying numpy call in such cases.
 
-    :param masks: Mask images in the [N, Height, Weight, 1] shape format.
-    :param img: Original base image
+    NOTE: Due to the batch nature of this function, utilizing a fill color will
+    consistently utilize more RAM due to the use of alpha blending
+
+    Assumptions:
+      * Mask input is per-pixel. Does not accept per-channel masks.
+      * Fill value input is in an applicable value range supported by the input
+        image format, which is mirrored in output images.
+
+    :param ref_image: Reference image to generate perturbations from.
+    :param masks: Mask matrix input of shape `[N x H x W]` where height and
+        width dimensions are the same size as the input `ref_image`.
+    :param fill: Optional fill color for the occluded regions as a single uint8
+        value or a per-channel sequence.
+
+    :raises ValueError: The input mask matrix was not 3-dimensional, its last
+        two dimensions did not match the shape of the input imagery, or the
+        input fill value could not be broadcast against the input image.
 
     :return: List of masked images in PIL Image form.
     """
-    masked_imgs = []
-    for mask in masks:
-        masked_img = np.multiply(mask, img, casting='unsafe')
-        masked_imgs.append(PIL.Image.fromarray(np.uint8(masked_img)))
-    return masked_imgs
+    if masks.ndim != 3:
+        raise ValueError(
+            f"Expected a 3-dimension mask input [N x H x W]. "
+            f"Got masks.ndim = {masks.ndim}"
+        )
+    if ref_image.size[::-1] != masks.shape[1:]:
+        raise ValueError(
+            f"Input image shape and mask image shape did not match: "
+            f"{ref_image.size[::-1]} != {masks.shape[1:]}"
+        )
+    ref_mode = ref_image.mode
+    ref_mat = np.asarray(ref_image)
+    s: Tuple = (...,)
+    if ref_mat.ndim > 2:
+        s = (..., None)  # add channel axis for multiplication
+    # The Batch Operation -> Bulk apply mask matrices to input image matrix to
+    # generate occluded images.
+    occ_img_mats = np.ndarray([len(masks), *ref_mat.shape], dtype=ref_mat.dtype)
+    if fill is not None:
+        masks_sview = masks[s]
+        np.add(
```

Comment:
I am curious what this was trying to achieve in the absence of a fill.

Reply:
You mean when `fill` *is* `None`?

Reply:
the use of the `UINT8_ONE` variable there (set to `np.uint8(1)` as a module constant) is to reduce the common  denominator when performing the `(1- mask)` computation. Numpy treats `1` as an `int64` type, causing a result matrix to use `8*H*W*C` bytes. The uint8 cast used causes the intermediate operation to yield a much smaller mask (`1*H*W*C` bytes) when mask is defined to be a smaller data type (like bool).